src/canvas/Document.tsx

```diff
@@ -1,19 +1,24 @@
 import { PatternPath } from './PatternPaths/PatternPath';
 import { Point } from './Geometry/Point';
 import { FaceFinder } from './Geometry/FaceFinder';
+import { PatternPathType } from './Enums';
+import { PatternPiece } from './PatternPiece';
+import { Vector } from './Geometry/Vector';
 
 export class Document implements IDocument {
     private _patternPaths: PatternPath[];
     private _patternPathsTrash: IPatternPathTrash[];
     private _sizeRatio: null | number; // in pixels per inch
-    private _patternPieces: PatternPath[][]; // an array of patternpath arrays, 
-                                             // each representing one piece of the pattern
+    private _patternPieces: PatternPiece[] | null;
+    private _allowanceSizes: Map<PatternPathType, number> | null; // allowance sizes in pixels
+    
 
     constructor () {
         this._patternPaths = new Array<PatternPath>();
         this._patternPathsTrash = [];
         this._sizeRatio = null;
-        this._patternPieces = [];
+        this._patternPieces = null;
```

Comment:
PatternPieces are another thing in document that cannot be set until we get a sizeRatio from the user, since PatternPieces contain allowances which require allowanceSizes which require the sizeRatio. So, like allowanceSizes, I thought it should remain null until it it possible to compute them.

Reply:
I see. I'm okay with that.